src/jit/treelifeupdater.cpp

```diff
@@ -0,0 +1,366 @@
+#include "jitpch.h"
+#ifdef _MSC_VER
+#pragma hdrstop
+#endif
+
+#include "treelifeupdater.h"
+
+template <bool ForCodeGen>
+TreeLifeUpdater<ForCodeGen>::TreeLifeUpdater(Compiler* compiler)
+    : compiler(compiler)
+    , newLife(VarSetOps::MakeEmpty(compiler))
+    , stackVarDeltaSet(VarSetOps::MakeEmpty(compiler))
+    , varDeltaSet(VarSetOps::MakeEmpty(compiler))
+    , gcTrkStkDeltaSet(VarSetOps::MakeEmpty(compiler))
+#ifdef DEBUG
+    , gcVarPtrSetNew(VarSetOps::MakeEmpty(compiler))
+    , epoch(compiler->GetCurLVEpoch())
+#endif // DEBUG
+{
+}
+
+//------------------------------------------------------------------------
+// UpdateLife: Determine does the tree affects liveness and processed accordingly.
+//
+// Arguments:
+//    tree - the tree which effect on liveness is processed.
+//
+template <bool ForCodeGen>
+void TreeLifeUpdater<ForCodeGen>::UpdateLife(GenTree* tree)
+{
+    assert(compiler->GetCurLVEpoch() == epoch);
```

Comment:
The new class is safer than the previous implementation because it saves epoch in the constructor and check that it doesn't change each time when we call its methods.